algviz/parsers/json_objects.py

```diff
@@ -0,0 +1,240 @@
+import json
+from . import structures
+
+class Tokens:
+    """Tokens we expect to see in the JSON"""
+    # First, keys we expect to see.
+    UID = "uid"
+    TYPE = "type"
+    FROM = "from"
+    TO = "to"
+    LEFT = "left"
+    RIGHT = "right"
+    DATA = "data"
+    GRAPH_NODES = "nodes"  # not a type, but an attribute of a graph
+    GRAPH_EDGES = "edges"
+    VARNAME = "var"
+    METADATA = "metadata"  # we probably should only use this for prototyping
+    TREE_ROOT = "root"
+    # Possible values for TYPE.  Keep these alphabetized and give them all the
+    # _T suffix, please.
+    ARRAY_T = "array"
+    BINARY_TREE_T = "bintree"
+    BINARY_TREE_NODE_T = "btnode"
+    EDGE_T = "edge"
+    GRAPH_T = "graph"
+    NODE_T = "node"
+    NULL_T = "null"
+    POINTER_T = "ptr"
+    STRING_T = "string"
+    WIDGET_T = "widget"
+
+aliases = {
+    # Shorter forms of tokens, to make it easier to handwrite these descriptions
+    "T": Tokens.TYPE,
+}
+
+_type_tokens = tuple(getattr(Tokens, name) for name in dir(Tokens)
+                     if name.endswith("_T") and not name.startswith("_"))
+_key_tokens = tuple(getattr(Tokens, name) for name in dir(Tokens)
+                    if not name.startswith("_") and not name.endswith("_T"))
+
+class JSONObjectError(Exception):
+    pass
+
+class Dispatcher(type):
+
+    @staticmethod
+    def dispatch(key):
+        # decorate an object
+        def _decorate(f):
+            f._purpose = key
+            return f
+        return _decorate
+
+    def __init__(cls, name, bases, namespace):
```

Comment:
shouldn't this be `__new__`?

Reply:
I don't see why.
You'll need to approve my changes before the branch can be merged into master, FYI.  You can also request further changes.

Reply:
By convention, cls is only used as the first argument for static methods such as new. init should have self as the first argument.

Reply:
This is a special case where the first object passed to the method is in fact the class whose metaclass is `Dispatcher`.  In particular, `cls` is not an instance of Dispatcher, so it's not appropriate to call it `self`.

Reply:
Classes with a specific metaclass are instances of that metaclass, see below:
![image](https://user-images.githubusercontent.com/32473809/33621676-935b8032-d9b9-11e7-91d8-adec3a35265b.png)
Anyway, as long as you added a comment for clarification, I don't think the point matters too much. I'll give your code another review later on today but it should be fine.

Reply:
I didn't know that.

Reply:
I like using `cls` anyway.  That's how I've seen `__init__` methods of metaclasses programmed in the past.  Also in the wild, in the first place I checked, which is here: https://eli.thegreenplace.net/2011/08/14/python-metaclasses-by-example/